### ISSUE TITLE
rustc incorrectly warn #[no_mangle] is not used when doing incremental build

### ISSUE THREAD

**Issue by wangbj** [rust-lang/rust#60050]
Steps to reproduce:

1) make an new crate, set crate type to `cdylib`
2) put a simple function in `src/lib.rs`:

```rust
#[no_mangle]
pub extern "C" fn rust_no_mangle() -> i32{
    42
}
```

3) build with `cargo build --verbose`
4) copy the `rustc` command line used from above step, and run it as-is

example:
```
$ cargo build --verbose
   Compiling nm v0.1.0 (/tmp/nm)
     Running `rustc --edition=2018 --crate-name nm src/lib.rs --color always --crate-type cdylib --emit=dep-info,link -C debuginfo=2 -C metadata=38c165a3ed3ed429 --out-dir /tmp/nm/target/debug/deps -C incremental=/tmp/nm/target/debug/incremental -L dependency=/tmp/nm/target/debug/deps`
    Finished dev [unoptimized + debuginfo] target(s) in 0.22s
$ rustc --edition=2018 --crate-name nm src/lib.rs --color always --crate-type cdylib --emit=dep-info,link -C debuginfo=2 -C metadata=38c165a3ed3ed429 --out-dir /tmp/nm/target/debug/deps -C incremental=/tmp/nm/target/debug/incremental -L dependency=/tmp/nm/target/debug/deps
warning: unused attribute
 --> src/lib.rs:1:1
  |
1 | #[no_mangle]
  | ^^^^^^^^^^^^
  |
  = note: #[warn(unused_attributes)] on by default

```

As you can see, the first time `cargo build` in step 3) has no warnings, however, step 4) cause `rustc` complains `#[no_mangle]` is unused, which isn't true.

Step 4) can be replaced by:

  * modify `src/lib.rs`, so that it cause rebuild
  * run `cargo build` again

rustc 1.35.0-nightly (70f130954 2019-04-16)

**Comment by MikailBag** [comment]
I have same error on today's nightly too

**Comment by bjorn3** [comment]
Same for me

**Comment by alexandermerritt** [comment]
I see the same warning text during a build for crate type “staticlib”.

**Comment by nagisa** [comment]
I will bisect this and then pass the baton to @oli-obk to fix this.

**Comment by nagisa** [comment]
Bisection claims 9217fe0e2f04d61dd29c9aaebee2c993705e1d26, (#59850). cc @Zoxc

**Comment by Zoxc** [comment]
My guess is that `no_mangle` should be a `Whitelisted` attribute? It was declared twice before https://github.com/rust-lang/rust/pull/59850 and the `Whitelisted` variant got removed.

**Comment by Zoxc** [comment]
Also our scheme to track unused attributes does not work the incremental compilation, which is probably why it only happens with that.

**Comment by Zoxc** [comment]
cc @petrochenkov

**Comment by oli-obk** [comment]
A fix is up. I this still nightly only or do we need to backport?

**Comment by jonas-schievink** [comment]
Seems to be nightly-only still

**Comment by elichai** [comment]
I still get this on nightly. `2019-05-20`

**Comment by oli-obk** [comment]
@elichai can you try with the latest nightly? This was only closed yesterday, so today's nightly should contain the fix.

**Comment by Disasm** [comment]
My case (https://github.com/rust-lang/rust/issues/60835) is no longer reproducible on 2019-05-21 nightly.

**Comment by phil-opp** [comment]
I no longer see the error with `rustc 1.36.0-nightly (50a0defd5 2019-05-21)`.

**Comment by elichai** [comment]
Yep it seems it's fixed. sorry I missed that it was merged yesterday. thanks!
